### PR TITLE
Add Multiple Balloons

### PR DESCRIPTION
**Pull request by RobertKeanu**

### CODE REVIEW

**Review by rob3rtu — APPROVED** [review]

File: Proiect3D/11_01_Shader.frag [hunk 1]
```diff
@@ -48,6 +48,78 @@ void main(void)
...
+    vec3 viewDir = normalize(inViewPos - FragPos);//vector catre observator normalizat (V)
+    vec3 reflectDir = reflect(-lightDir, normala); // reflexia razei de lumina (R)
+    float spec = pow(max(dot(viewDir, reflectDir), 0.0), 1);
+    vec3 specular = specularStrength * spec * lightColor;  
+    vec3 emission=vec3(0.0, 0.0, 0.0);
+    vec3 result = emission+(ambient + diffuse + specular) * vec3(0.0f, 0.0f, 1.0f);
+	out_Color = vec4(result, 1.0f);
+    }
+
+     if (codCol==4) // galben
```
Comment (andlft) [inline Proiect3D/11_01_Shader.frag:61]: There is too much duplicate code here, use the if statement only to set the color.

File: Proiect3D/11_01_Shader.vert [hunk 2]
```diff
@@ -41,6 +41,37 @@ void main(void)
...
+    {
+		gl_Position = projection*view*myMatrix*in_Position;
+        Normal =mat3(projection*view*myMatrix)*in_Normal; 
+        inLightPos = vec3(projection*view*myMatrix* vec4(lightPos, 1.0f));
+        inViewPos =vec3(projection*view*myMatrix*vec4(viewPos, 1.0f));
+        dir = mat3(projection*view*myMatrix) * vec3(0.0,100.0,200.0); // pentru sursa directionala
+        FragPos = vec3(gl_Position);
+    }
+
+    if (codCol==4)
```
Comment (andlft) [inline Proiect3D/11_01_Shader.vert]: Why do we need multiple if statements to do the same thing?

File: Proiect3D/11_01_umbra.cpp [hunk 3]
```diff
@@ -299,11 +305,28 @@ void DestroyShaders(void)
 	glDeleteProgram(ProgramId);
 }
 
+int getRandomNumber(const std::vector<int>& numbers) {
+	// Generate a random index to choose a number from the vector
+	int randomIndex = std::rand() % numbers.size();
+
+	// Return the randomly chosen number
```
Comment (andlft) [inline Proiect3D/11_01_umbra.cpp]: Is this generated by ChatGPT?

File: Proiect3D/11_01_umbra.cpp [hunk 4]
```diff
@@ -316,61 +339,34 @@ void Initialize(void)
...
-	glUniform3f(lightColorLocation, 1.0f, 1.0f, 1.0f);
-	glUniform3f(lightPosLocation, xL, yL, zL);
-	glUniform3f(viewPosLocation, Obsx, Obsy, Obsz);
-
+	glUniformMatrix4fv(myMatrixLocation, 1, GL_FALSE, &myMatrix[0][0]);
 	// desenare cub
 	codCol = 0;
 	glUniform1i(codColLocation, codCol);
-	myMatrix = glm::mat4(1.0f);
+	myMatrix = mymatrix;
```
Comment (andlft) [inline Proiect3D/11_01_umbra.cpp]: you should rename `mymatrix`, the name is too similar.

File: Proiect3D/11_01_umbra.cpp [hunk 5]
```diff
@@ -413,6 +409,51 @@ void RenderFunction(void)
...
+	glUniform3f(lightPosLocation, xL, yL, zL);
+	glUniform3f(viewPosLocation, Obsx, Obsy, Obsz);
+
+	myMatrix = glm::mat4(1.0f);
+	DrawBalloon(myMatrix,0);
+	//vector random pt x,y
+	l = l + 0.8;
+	for (int i = 1; i <= 20; i++)
+	{
+		matrTransl = glm::translate(glm::mat4(1.0f), glm::vec3(xpos[i], ypos[i], std::fmod((l + i * 120), 5000.0f)));
```
Comment (andlft) [inline Proiect3D/11_01_umbra.cpp]: I don't think an offset of 120 is enough, some ballons might still overlap. There shouldn't be multiple balloons on the same height level.
image

File: Proiect3D/11_01_umbra.cpp [hunk 6]
```diff
@@ -413,6 +409,51 @@ void RenderFunction(void)
...
+		matrRot = glm::rotate(glm::mat4(1.0f), PI / 4, glm::vec3(0.0, 0.0, 1.0));
```
Comment (andlft) [inline Proiect3D/11_01_umbra.cpp]: Remove this rotation.

**Review by andlft — CHANGES_REQUESTED** [review]
Adjust parameters so balloons spawn regularly, there are many balloons that spawn at the beginning of the scene and after that there is a pause of about 15 seconds before balloons start spawning again.

**Review by andlft — APPROVED** [review]

**Pull request merged**